### ISSUE TITLE
General Purpose Data Driven Finite State Machine Plugin

### ISSUE THREAD

**Issue by PradeepKumarRajamanickam** [bevyengine/bevy#629]
It would be nice to have a high performance data driven fsm with an easy to use visual editor similar to Ludiq Bolt on Unity. It will move various kinds of control flow out of code and into data. 

**Benefits**
- It can be used to build a number of other content authoring systems on top of it.
eg, Visual Scripting, Animation State Machine, AI FSM
- Reduces entry level knowledge requirement to make games due to visual nature of the tool
- Content creation (UX, AI...etc) can be distributed to non-programming staff
- Makes large scale projects (AA & AAA games) easier to manage
- Reduced compile time on large projects

[Bolt Unity](https://assetstore.unity.com/packages/tools/visual-scripting/bolt-163802)

**Comment by PradeepKumarRajamanickam** [comment]
I think I might take a shot at it once I wrap up input map plugin development.

**Comment by walterpie** [comment]
Would the "general purpose" aspect of this allow the engine to control the scheduler, i.e. conditionally execute systems? Or is this purely for game logic? If the former, I would love to help on the scheduler aspect of things (and really anything that has to do with this)

**Comment by PradeepKumarRajamanickam** [comment]
> Would the "general purpose" aspect of this allow the engine to control the scheduler, i.e. conditionally execute systems? Or is this purely for game logic? If the former, I would love to help on the scheduler aspect of things (and really anything that has to do with this)

Yes in order for this to be user friendly or even possible, we need the ability to add and remove a bunch systems based on the current state. I am really interested in the idea of providing these api's in Command as mentioned in this discussion https://github.com/bevyengine/bevy/discussions/654#discussioncomment-95071. This will make things a lot easier.

This tool/system is mainly for content authoring high-level game/app related stuff eg, ux flow, gameplay, ai, animation fsm...etc. I believe using visual tools for these kinds of high-level game-specific content make actual projects with complex flow much easier to maintain.

I will be penning down my thoughts and share them with the community. To see if there are any technical red flags, regarding impl approach that may not be possible. But this is gonna be a while before it happens.

**Comment by PradeepKumarRajamanickam** [comment]
I am working on an idea.

Brahma Overview
https://hackmd.io/@-EznnAUiQTygOuhjfvLeMg/SkgCxkQwv

**Comment by zicklag** [comment]
@PradeepKumarRajamanickam I was just thinking that it would be great if we could abstract the FSM and other logic into a crate that could produce some intermediate representation that would be translated to Rust code in a separate stage. Also it would be great to keep the logic separate from the UI. This would abstract both the input ( the UI ) and the output ( the code) so that they could be swapable.

My primary motivation for this is that this kind of functionality is something that I would love to integrate into [Arsenal](https://github.com/katharostech/arsenal). In Arsenal we would be using Blender's node system to power the UI and we would compile not to Rust but to a scripting language, probably either Lua or Python. Therefore we would have a different frontend GUI and a different backend code target, but everything else would probably function the same.

**Comment by PradeepKumarRajamanickam** [comment]
@zicklag 
"we could abstract the FSM and other logic into a crate that could produce some intermediate representation that would be translated to Rust code in a separate stage". 
I was wondering the same thing. Using ron files for graph persistence and generating *.rs src only when the user explicitly hits the "compile" button would be better.

" Also it would be great to keep the logic separate from the UI"
Yes you are right there should be a separate glue layer on top

I have updated the diagram have a look.

**Comment by PradeepKumarRajamanickam** [comment]
Update*
- Started work on brahma_graph 
https://github.com/PradeepKumarRajamanickam/brahma_graph/tree/pradeep/dev
- Added more info to overview doc
https://hackmd.io/@-EznnAUiQTygOuhjfvLeMg/SkgCxkQwv

**Comment by Weibye** [comment]
> It would be nice to have a robust, high performance data driven fsm with an easy to use visual editor similar to Ludiq Bolt on Unity. It will move various kinds of control flows out of code and into data.

I think before a fancy visual node-editor statemachine is implemented (I really _would_ want that) I believe Bevy needs a working Finite State Machine implementation that can serve as the building block for many other features. Then later that can be extended to incorporate visual editors.

What is the current status of this @PradeepKumarRajamanickam?
